src/transformers/modeling_blenderbot.py

```diff
@@ -156,6 +164,23 @@ def prepare_inputs_for_generation(self, decoder_input_ids, past, attention_mask,
             "attention_mask": attention_mask,
             "use_cache": use_cache, 
         }
+    
+    def get_input_embeddings(self):
+        return self.shared
+
+    def set_input_embeddings(self, value):
+        self.shared = value
+        self.encoder.embed_tokens = self.shared
+        self.decoder.embed_tokens = self.shared
+
+    def get_output_embeddings(self):
```

Comment:
This function has to be changed. We don't want to instantiate a new ``nn.Linear()` every time this function is called (at every `generate()` call for example. Why don't you just return `self.shared` here?

Reply:
I see `modeling_bart` does the same...I don't really understand why we need to do this... => why not just return self.shared? Also cc @sshleifer

Reply:
[test_modeling_common.py](https://github.com/huggingface/transformers/blob/master/tests/test_modeling_common.py#L573) requires `nn.Linear`, not `nn.Embedding`. We should type hint.


```python
            x = model.get_output_embeddings()
            self.assertTrue(x is None or isinstance(x, torch.nn.Linear))
```

Reply:
Yes exactly that's why I added it here. same for `get_encoder`

Reply:
Ok, but I don't think this is the correct way to do it here (see my comment below). We can instantiate a layer one time in the `__init__()` this will automatically be tied to the input embeddings => so no need to train this layer. Taking a look into the `init_weights()` function in modeling_utils.py should make this clearer. I think we should change this both in Bart and here.

Reply:
What is the evidence that this layer gets trained? I never call this function. I call
```python
lm_logits = F.linear(outputs[0], self.model.shared.weight) 
```
on L972 of bart.

I didn't want to instantiate `lm_head` because it wastes 200MB for no reason.

Reply:
Sorry, maybe that wasn't clear here. 
I meant that since a lot of models share the input and output embeddings, a layer `lm_head` is added to the `__init__()`  but then overwritten by the weights of the input embeddings when the models calls `init_weights`, see here: https://github.com/huggingface/transformers/blob/bc3a0c06075050d3de586c543e4ad6a7efc9260e/src/transformers/modeling_utils.py#L357

This means that while the layer is instantiated at __init__ the weights are overwritten, garbage collector frees the memory, so that no 200MB are wasted. `tie_weights` does exactly the same as you do in the `get_output_embeddings()` function, but only once at __init__, IMO, there is no advantage to creating `nn.Linear` on the fly in `get_output_embeddings()`: 
Here my arguments:

- Adding a `nn.Linear` to `__init__` means that memory is shortly increased by allocating memory for the weights, the weights are overwritten so that this memory is released directly afterward (The exact same thing happens when `get_ouput_embeddings()` is called in Bart at the moment). IMO, it's much better to do this only once when instantiating the model instead of doing this every time `get_output_embeddings()` is called.
- As a user, I would not expect the function `get_ouput_embeddings()` to generate a  `nn.Linear` on the fly, since it's a getter method. This method is actually quite expensive memory-wise, which I don't think a getter method should be. Also `get_output_embeddings` is called every time when calling `generate()` 
- The design is not consistent with other models. The `tie_or_clone_weights()` method was implemented exactly for the reason that input and output embeddings share the same weights and is used for all other models. IMO, it's confusing to see a different design pattern at Bart when adding a model as it's done here. 
- (Not 100p about this) I think for torchscript and for an automatic TF conversion function, the model has to have a `nn.Linear` parameter with a name in the model class. 

All in all, I really think we should change that. I don't see an advantage of this design choice.

Reply:
OK, I will profile it and move it to init if it doesn't waste mem. Expect to see a PR today or tomorrow.
@mariamabarham you can keep going with this and I will fix both if you merge before me :) Does that work @patrickvonplaten ?

Reply:
Sounds great!